### PR TITLE
Handling external deletion of the resources in the clusterresources group

### LINKED ISSUE #572: External Deletion
When a resource has been deleted using any method other than the k8s operator, our task is to prevent the deletion of the resource within the Kubernetes environment. Instead, we aim to retain the resource and provide a notification indicating that it has been removed from the Instaclustr

### PR DESCRIPTION
**Pull request by worryg0d**
# Issue 572

Added:
- handling of the external deletion of the resources on Instaclustr
- fixed external changes for VPC peering resources when Instaclustr API returns a list of subnets in a different order

closes #572

### CODE REVIEW

File: pkg/instaclustr/mock/client.go [hunk 1]
```diff
@@ -375,3 +375,7 @@ func (c *mockClient) GetResizeOperationsByClusterDataCentreID(cdcID string) ([]*
 func (c *mockClient) GetAWSVPCPeering(peerID string) (*models.AWSVPCPeering, error) {
 	panic("GetAWSVPCPeering: is not implemented")
 }
+
+func (c *mockClient) CheckIfAWSEndpointServicePrincipalExists(id string) (bool, error) {
+	panic("CheckIfAWSEndpointServicePrincipalExists: is not implemented")
```
Comment (OleksiienkoMykyta) [inline pkg/instaclustr/mock/client.go:380]: Everything is great, but I think we should cover it with tests in future

**Review by OleksiienkoMykyta — APPROVED** [review]

**Comment by OleksiienkoMykyta** [comment]
I agree with your statment, we should rename "state" to "instaclustrResourceState"

**Review by tengu-alt — APPROVED** [review]

File: pkg/instaclustr/client.go [hunk 2]
```diff
@@ -2215,6 +2215,31 @@ func (c *Client) UpdateClusterSettings(clusterID string, settings *models.Cluste
 	return nil
 }
 
+func (c *Client) CheckIfAWSEndpointServicePrincipalExists(id string) (bool, error) {
+	url := c.serverHostname + AWSEndpointServicePrincipalEndpoint + "/" + id
```
Comment (testisnullus) [inline pkg/instaclustr/client.go:2219]: `AWSEndpointServicePrincipalEndpoint` already has a backslash at the end, why did you add an additional one?

File: pkg/instaclustr/client.go [hunk 3]
```diff
@@ -2215,6 +2215,31 @@ func (c *Client) UpdateClusterSettings(clusterID string, settings *models.Cluste
 	return nil
 }
 
+func (c *Client) CheckIfAWSEndpointServicePrincipalExists(id string) (bool, error) {
```
Comment (testisnullus) [inline pkg/instaclustr/client.go:2218]: Let's rename this method to GetAWSEndpointServicePrincipal and will return the service principal entity itself and an error. We can check the existence of the principal by using `errors.Is(err, instaclustr.NotFound)` func also

File: controllers/clusterresources/awsvpcpeering_controller.go [hunk 4]
```diff
@@ -49,6 +48,8 @@ type AWSVPCPeeringReconciler struct {
 	EventRecorder record.EventRecorder
 }
 
+const awsVPCPeeringStatusCodeDeleted = "deleted"
```
Comment (testisnullus) [inline controllers/clusterresources/awsvpcpeering_controller.go:51]: Let's use const from package models instead of this variable

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: pkg/instaclustr/client.go (hunk 3 above)
Comment (worryg0d) [reply pkg/instaclustr/client.go]: Fixed

File: controllers/clusterresources/awsvpcpeering_controller.go (hunk 4 above)
Comment (worryg0d) [reply controllers/clusterresources/awsvpcpeering_controller.go]: Moved to the models pkg

**Review by testisnullus — APPROVED** [review]

**Comment by ribaraka** [comment]
> I want to hear your opinion about the added state field which indicates the state of the resource on Instaclustr. Maybe changing it to something like `instaclustrResourceState` will be better.
> 

The source of truth is the Instaclustr Console. So whatever the state/status of the Instaclustr resource is, it should be directly reflected on the k8s `state` field without any deviation. For now, I don't see a good reason to create a new field or give a new name to the existing `state` field of the resource.

> Because some resources also have their status codes like AWS VPC peering. So it looks a bit weird when there is the following status:
> 
> 
> ```
> status:
>   state: CREATED
>   statusCode: deleted
> 
> ```
> 
> The field state indicates that the resource is created on Instaclustr, but statusCode: deleted indicates that the VPC peering is deleted on AWS.
> 

In this example, I suggest you maintain consistency between the `state` and `statusCode` fields, or, even better, you can consider using only the `state` field to represent the `statusCode`.

File: controllers/clusterresources/awsencryptionkey_controller.go [hunk 5]
```diff
@@ -129,6 +129,7 @@ func (r *AWSEncryptionKeyReconciler) handleCreate(
 		)
 
 		encryptionKey.Status = *encryptionKeyStatus
+		encryptionKey.Status.State = models.CreatedStatus
```
Comment (ribaraka) [inline controllers/clusterresources/awsencryptionkey_controller.go:132]: try the thing I suggested by replacing `statusCode` with `state` or vice versa, as you prefer.

File: controllers/clusterresources/awsencryptionkey_controller.go [hunk 6]
```diff
@@ -296,6 +303,24 @@ func (r *AWSEncryptionKeyReconciler) newWatchStatusJob(encryptionKey *v1beta1.AW
 	}
 }
 
+func (r *AWSEncryptionKeyReconciler) handleExternalDelete(ctx context.Context, key *v1beta1.AWSEncryptionKey) error {
+	l := log.FromContext(ctx)
+
+	patch := key.NewPatch()
+	key.Status.State = models.DeletedStatus
```
Comment (ribaraka) [inline controllers/clusterresources/awsencryptionkey_controller.go:310]: again, the thing with displaying the `statuscode` in the `state`

File: controllers/clusterresources/awsendpointserviceprincipal_controller.go [hunk 7]
```diff
@@ -175,6 +188,44 @@ func (r *AWSEndpointServicePrincipalReconciler) handleDelete(ctx context.Context
...
+			if errors.Is(err, instaclustr.NotFound) {
+				return r.handleExternalDelete(ctx, principal)
+			}
+
+			return err
+		}
+
+		return nil
+	}
+}
```
Comment (ribaraka) [inline controllers/clusterresources/awsendpointserviceprincipal_controller.go:209]: add a checker here for the existence of k8s resource

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/awsencryptionkey_controller.go (hunk 5 above)
Comment (worryg0d) [reply controllers/clusterresources/awsencryptionkey_controller.go:132]: Started using existing `Status` field

File: controllers/clusterresources/awsencryptionkey_controller.go (hunk 6 above)
Comment (worryg0d) [reply controllers/clusterresources/awsencryptionkey_controller.go:326]: fixed

File: controllers/clusterresources/awsendpointserviceprincipal_controller.go (hunk 7 above)
Comment (worryg0d) [reply controllers/clusterresources/awsendpointserviceprincipal_controller.go]: Added here and for other resources

**Review by ribaraka — APPROVED** [review]
